kolena/_experimental/dataset/common.py

```diff
@@ -45,14 +45,19 @@ def validate_id_fields(id_fields: List[str], existing_id_fields: List[str] = Non
             )
 
 
+def _convert_non_int_str_to_str(df: pd.DataFrame) -> pd.DataFrame:
+    df = df.applymap(lambda x: str(x) if (not isinstance(x, (int, str))) else x)
```

Comment:
I think this would have unexpected result when `str(x)` is not applied on trivial objects, for example class instances. If we choose to restrict id column type, I'd prefer not to auto convert but be explicit.

Thinking about this again, I'm retracting my initial suggestion and think we probably don't need this enforcement on int or str type. I think the only "problematic" part is floating number comparison, and we have our existing precision truncation on server side for that. What do you think?

Reply:
- ingestion via web UI
  - only basic types via parsing through json
- ingestion via python script
  - it might fail for non-trivial class, but users can convert it into a dict

i think this change will make the system more robust. currently it will just throw some pandas error when selecting dict and other type as id fields as unit tests.

i think we want to have this uniqueness check on the client side no matter what dedup we have on server side

Reply:
I want to highlight the problem of using `str` as auto type coercion. Using the new unit test, this example would pass, but it should fail. (there could be others, as `str` is not designed for id)

```
        (pd.DataFrame(dict(a=[{"a": 1, "b": 2}, {"a": 2, "b": 1}, {"b": 2, "a": 1}], b=[1, 2, 1])), ["a", "b"]),
```

Reply:
synced offline, string coercion is not the right approach to compare uniqueness, [converted the multiple columns df into a single column](https://github.com/kolenaIO/kolena/pull/368/commits/c1e855c3212633a888649edb12c6d0e8902a8efc) and leverage the default dictionary comparison might do the trick